pkg/gc/gc_state_manager_test.go

```diff
@@ -849,6 +876,247 @@ func (s *gcStateManagerTestSuite) TestGCBarriers() {
+
+	b, err = s.manager.DeleteGlobalGCBarrier(context.Background(), "b2")
+	re.NoError(err)
+	expected = endpoint.NewGlobalGCBarrier("b2", 27, ptime(now.Add(time.Hour)))
+	re.Equal(expected, b)
+	re.Empty(s.getGlobalGCBarriers())
+
+	res, err = s.manager.AdvanceTxnSafePoint(keyspaceID, 30, now)
+	re.NoError(err)
+	re.Equal(uint64(27), res.OldTxnSafePoint)
+	re.Equal(uint64(30), res.NewTxnSafePoint)
+	re.Equal(uint64(30), res.Target)
+	re.Empty(res.BlockerDescription)
+
+	// Deleting non-existing GC barrier.
+	b, err = s.manager.DeleteGlobalGCBarrier(context.Background(), "b1")
+	re.NoError(err)
+	re.Nil(b)
+
+	// Test TTL
+	_, err = s.manager.SetGlobalGCBarrier(context.Background(), "b3", 40, time.Minute, now)
+	re.NoError(err)
+	_, err = s.manager.SetGlobalGCBarrier(context.Background(), "b4", 45, time.Minute*2, now)
+	re.NoError(err)
+	_, err = s.manager.SetGlobalGCBarrier(context.Background(), "b5", 50, time.Duration(math.MaxInt64), now)
+	re.NoError(err)
+
+	// Not expiring
+	for _, t := range []time.Time{now, now.Add(time.Second * 59), now.Add(time.Minute)} {
+		res, err = s.manager.AdvanceTxnSafePoint(keyspaceID, 60, t)
+		re.NoError(err)
+		re.Equal(uint64(40), res.NewTxnSafePoint)
+		re.Contains(res.BlockerDescription, "BarrierID: \"b3\"")
+		s.checkTxnSafePoint(keyspaceID, 40)
+		re.Len(s.getGlobalGCBarriers(), 3)
+	}
+
+	// b3 expires
+	res, err = s.manager.AdvanceTxnSafePoint(keyspaceID, 60, now.Add(time.Minute*2))
+	re.NoError(err)
+	re.Equal(uint64(45), res.NewTxnSafePoint)
+	re.Contains(res.BlockerDescription, "BarrierID: \"b4\"")
+	s.checkTxnSafePoint(keyspaceID, 45)
+	re.Len(s.getGlobalGCBarriers(), 2)
+
+	// b4 expires, but b5 never expires.
+	res, err = s.manager.AdvanceTxnSafePoint(keyspaceID, 60, now.Add(time.Hour*24*365*100))
+	re.NoError(err)
+	re.Equal(uint64(50), res.NewTxnSafePoint)
+	re.Contains(res.BlockerDescription, "BarrierID: \"b5\"")
+	s.checkTxnSafePoint(keyspaceID, 50)
+	re.Len(s.getGlobalGCBarriers(), 1)
+
+	// Manually delete b5
+	b, err = s.manager.DeleteGlobalGCBarrier(context.Background(), "b5")
+	re.NoError(err)
+	re.Equal("b5", b.BarrierID)
+
+	res, err = s.manager.AdvanceTxnSafePoint(keyspaceID, 60, now.Add(time.Hour*24*365*100))
+	re.NoError(err)
+	re.Equal(uint64(60), res.NewTxnSafePoint)
+	s.checkTxnSafePoint(keyspaceID, 60)
+
+	re.Empty(s.getGlobalGCBarriers())
+
+	// Disallows setting GC barrier before txn safe point.
+	_, err = s.manager.SetGlobalGCBarrier(context.Background(), "b6", 50, time.Hour, now)
+	re.Error(err)
+	re.ErrorIs(err, errs.ErrGlobalGCBarrierTSBehindTxnSafePoint)
+	re.Empty(s.getGlobalGCBarriers())
+	// BarrierTS exactly equals to txn safe point is allowed.
+	b, err = s.manager.SetGlobalGCBarrier(context.Background(), "b6", 60, time.Hour, now)
+	re.NoError(err)
+	expected = endpoint.NewGlobalGCBarrier("b6", 60, ptime(now.Add(time.Hour)))
+	re.Equal(expected, b)
+	re.Len(s.getGlobalGCBarriers(), 1)
+	re.Equal(expected, s.getGlobalGCBarrier("b6"))
```

Comment:
Multi-keyspace cases are not covered. Consider cover the case that multiple keyspaces (including the null keyspace) has different txn safe points.

Reply:
done